external/llvm-project/mlir/lib/Conversion/XModelToGPU/XModelToGPU.cpp

```diff
@@ -91,45 +91,74 @@ struct LaunchRewritePattern : public OpRewritePattern<async::LaunchOp> {
     return b.create<gpu::WaitOp>(loc, tokenType, deps).asyncToken();
   }
 
-  Value moveMemory(OpBuilder b, Value opr, uint32_t fidx, bool readAccess,
-                   bool writeAccess, llvm::SmallVector<Value> &copyBackOprs,
+  bool isOnDevice(ArrayRef<Operation *> oprUsers) const {
+    for (auto opUse : oprUsers) {
+      auto launch = dyn_cast<async::LaunchOp>(opUse);
+      // assumes the same GPU
+      if (!launch || !getGPUTarget(launch).has_value())
+        return false;
+    }
+    return true;
+  }
+
+  Value moveMemory(OpBuilder b, async::LaunchOp launchOp, Value opr,
+                   uint32_t fidx, bool readAccess, bool writeAccess,
+                   llvm::SmallVector<Value> &copyBackOprs,
                    llvm::SmallVector<Value, 8> &asyncDeps) const {
     Location loc = opr.getLoc();
     auto tokenType = b.getType<gpu::AsyncTokenType>();
     auto oprAllocOp = opr.getDefiningOp<memref::AllocOp>();
+    auto bAlloc = b;
```

Comment:
I think this could be replaced with an insertion point guard? I don't know how expensive the builder copy constructor is

Reply:
The memcpy needs to be inserted at 'b' location, whereas the alloc may be earlier.

Reply:
Yes

So put the bits that need to go earlier inside an extra set of curly braces, summon up an insertion point guard, and then `setInsertionPoint()` on b? Or did I miss a use of `b` vs `bAlloc` that prevents that structure?